### DISCUSSION TITLE
Entering and exiting state

### DISCUSSION THREAD

**Discussion by lukors** [bevyengine/bevy#3918] [Q&A] [answered]
Is it possible to enter/exit a state right away?

So let's say I'm in `MyState::StateTwo`, and I want to exit and enter the state. I was hoping I could do this:

```rust
my_state.push(MyState::StateOne);
my_state.push(MyState::StateTwo);
```

But seems like it can't have more than one state queued at a time?

What I'm actually trying to accomplish is to run the `with_run_criteria(State::on_enter(MyState::StateTwo))` functions again if the state is already active, and would prefer to not have to add special code for this.

**Comment by MrGVSV** [comment, marked as answer]
My PR (#3328) regarding this issue was recently merged. That's probably what you need.

**Reply by lukors** [reply]
Yeah that would make my case significantly better! :)

I could do something like this:

```rust
if let StateError::AlreadyInState = my_state.overwrite_replace(MyState::StateTwo) {
    my_state.overwrite_restart() 
}
```

It would be even better with something like a `my_state.push_or_restart();` that does this in one function call, but I can do without. :)

Thanks! :D
